File: HW06_SNS_SQS/Node/src/sqsConsumer.js
require('dotenv').config();

var AWS = require('aws-sdk');
AWS.config.update({ region: `${process.env.REGION}` });

const MAX_MESSAGES = 1000;
let msgCounter = 0;
let oddCounter = 0;
let evenCounter = 0;

var sqs = new AWS.SQS({ apiVersion: '2012-11-05' });
var queueURL = `${process.env.QUEUE_URL}`;

var params = {
  AttributeNames: [
    "SentTimestamp"
  ],
  MaxNumberOfMessages: 10,
  MessageAttributeNames: [
    "All"
  ],
  QueueUrl: queueURL,
  VisibilityTimeout: 20,
  WaitTimeSeconds: 0
};

var consume = function() {
  sqs.receiveMessage(params, function(err, data) {
    if (err) {
      console.log("Receive Error", err);
    }
    if (data.Messages) {
      for (var i = 0; i < data.Messages.length; i++) {
        var message = data.Messages[i];
        var number = JSON.parse(message.Body);
        
        msgCounter++;

        let isOdd = number % 2 == 0 ? false : true;
        if (isOdd) {
          oddCounter++;
        } else {
          evenCounter++;
        }

        console.log("Message " + msgCounter + ". " + number + ". oddConter = " + oddCounter + ", evenCounter = " + evenCounter);

        removeFromQueue(message);
      }
      consume();
    } else {
      setTimeout(function() {
        receiveMessage()
      }, 60 * 1000);
    }
  });
};

var removeFromQueue = function(message) {
  sqs.deleteMessage({
      QueueUrl : queueURL,
      ReceiptHandle : message.ReceiptHandle
  }, function(err, data) {
      err && console.log(err);
  });
};

consume();

File: HW06_SNS_SQS/Python/src/SQSConsumer.py
import boto3
import time

boto3.setup_default_session(profile_name='YOUR_PROFILE_NAME')

MAX_MESSAGES = 1000
msgCounter = 0
oddCounter = 0
evenCounter = 0

sqs = boto3.client('sqs', region_name='YOUR_REGION')
queue_url = 'https://sqs.YOUR_REGION.amazonaws.com/YOUR_QUEUE_URL'

while msgCounter < MAX_MESSAGES:
    response = sqs.receive_message(
        QueueUrl=queue_url,
        AttributeNames=[
            'SentTimestamp'
        ],
        MaxNumberOfMessages=1,
        MessageAttributeNames=[
            'All'
        ],
        VisibilityTimeout=0,
        WaitTimeSeconds=0
    )

    try:
        message = response['Messages'][0]
        receipt_handle = message['ReceiptHandle']

        sqs.delete_message(
            QueueUrl=queue_url,
            ReceiptHandle=receipt_handle
        )

        number = int(message['Body'])

        msgCounter += 1

        if (number % 2) == 0:
            evenCounter += 1
        else:
            oddCounter += 1

        print("Message " + str(msgCounter) + ". " + str(number) + ". oddCounter = " + str(oddCounter) + ", evenCounter = " + str(evenCounter))

    except KeyError:
        print('No Messages on the Queue')
        message = []
        time.sleep(60)
